### PR TITLE
Ncsdk 9897 caf pm

### PR DESCRIPTION
**Pull request by rakons**
This PR moves the power_manager from pelion_client into CAF and uses in inside nrf_desktop.

### CODE REVIEW

File: applications/nrf_desktop/src/modules/hid_state_pm.c
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_state_pm.c:2]: date needs update

File: applications/nrf_desktop/src/modules/hid_state_pm.c
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_state_pm.c:16]: neat :)

File: applications/nrf_desktop/src/modules/usb_state_pm.c
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/usb_state_pm.c:34]: excessive empty line

File: applications/nrf_desktop/src/modules/usb_state_pm.c
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/usb_state_pm.c:2]: date - check other new files

File: include/caf/events/force_power_down_event.h
Comment (pdunaj) [inline include/caf/events/force_power_down_event.h:31]: power

File: include/caf/events/force_power_down_event.h
Comment (pdunaj) [inline include/caf/events/force_power_down_event.h:1]: Wondering if it won't be better to keep power manager related events in a single file. Also we could think about some common prefix for them.

Anyway - this can be done later as cleanup while documentation is added,

File: subsys/caf/modules/ble_state_pm.c
Comment (pdunaj) [inline subsys/caf/modules/ble_state_pm.c:16]: move to tightest scope (inside function - line 21)

File: subsys/caf/modules/ble_state_pm.c
Comment (pdunaj) [inline subsys/caf/modules/ble_state_pm.c:44]: excessive empty line (or add line after each break)

File: subsys/caf/modules/power_manager.c
Comment (pdunaj) [inline subsys/caf/modules/power_manager.c:327]: if keep alive event can be enabled/disabled we should check it here too

**Review by pdunaj — CHANGES_REQUESTED** [review]
Looks ok. Small tweaks needed.

File: applications/nrf_desktop/src/events/usb_event.h
Comment (rakons) [reply applications/nrf_desktop/src/events/usb_event.h:20]: ok

File: applications/nrf_desktop/src/modules/hid_state_pm.c
Comment (rakons) [reply applications/nrf_desktop/src/modules/hid_state_pm.c:2]: ok

File: applications/nrf_desktop/src/modules/hid_state_pm.c
Comment (rakons) [reply applications/nrf_desktop/src/modules/hid_state_pm.c:16]: ;)

File: applications/nrf_desktop/src/modules/usb_state_pm.c
Comment (rakons) [reply applications/nrf_desktop/src/modules/usb_state_pm.c:34]: ok

File: applications/nrf_desktop/src/modules/usb_state_pm.c
Comment (rakons) [reply applications/nrf_desktop/src/modules/usb_state_pm.c:2]: ok

File: include/caf/events/force_power_down_event.h
Comment (rakons) [reply include/caf/events/force_power_down_event.h:31]: ok

File: include/caf/events/force_power_down_event.h
Comment (rakons) [reply include/caf/events/force_power_down_event.h:1]: In my feeling it is better to split it as it makes it easier to trace the code and capsule it.

File: subsys/caf/modules/ble_state_pm.c
Comment (rakons) [reply subsys/caf/modules/ble_state_pm.c:16]: ok

File: subsys/caf/modules/ble_state_pm.c
Comment (rakons) [reply subsys/caf/modules/ble_state_pm.c:44]: I prefer to limit number of useless empty lines :)

File: subsys/caf/modules/power_manager.c
Comment (rakons) [reply subsys/caf/modules/power_manager.c:327]: Done

File: applications/nrf_desktop/src/modules/usb_state_pm.c
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/usb_state_pm.c:44]: If there will be something to do, these returns should be switched to breaks and one return left after switch case.

File: include/caf/events/force_power_down_event.h
Comment (pdunaj) [inline include/caf/events/force_power_down_event.h:16]: Simple event to force a quick power down

**Review by pdunaj — APPROVED** [review]
small changes needed in doc text
one suggestion
otherwise looks ok

File: applications/nrf_desktop/src/modules/usb_state_pm.c
Comment (rakons) [reply applications/nrf_desktop/src/modules/usb_state_pm.c]: fixed

File: include/caf/events/force_power_down_event.h
Comment (rakons) [reply include/caf/events/force_power_down_event.h]: fixed

**Review by pdunaj — APPROVED** [review]

File: include/caf/power_manager.rst
Comment (MarekPieta) [inline include/caf/power_manager.rst:3]: `CAF: Power manager module`

File: include/caf/power_manager.rst
Comment (MarekPieta) [inline include/caf/power_manager.rst:1]: We will have to align documentation with other CAF modules and rephrase notes that are specific to nRF Desktop.
I think we should do it in separate PR (this one already introduces quite a lot of changes).

File: subsys/caf/modules/power_manager.c
Comment (MarekPieta) [inline subsys/caf/modules/power_manager.c:328]: CONFIG_CAF_KEEP_ALIVE_EVENTS is actually selected by the CONFIG_CAF_POWER_MANAGER. Do we need the #if here ?

File: subsys/caf/modules/Kconfig.ble_state
Comment (MarekPieta) [inline subsys/caf/modules/Kconfig.ble_state:26]: we have

File: include/caf/events/force_power_down_event.h
Comment (MarekPieta) [inline include/caf/events/force_power_down_event.h:16]: We should add documentation for newly added events into generated rst files (I think that could be separate PR)

File: subsys/caf/events/force_power_down_event.c
Comment (MarekPieta) [inline subsys/caf/events/force_power_down_event.c:25]: CAF_INIT_LOG_FORCE_POWER_DOWN_EVENTS

File: applications/nrf_desktop/src/modules/hid_state_pm.c
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/hid_state_pm.c:18]: This adds additional event handling to our critical path. We may need to verify if this has a negative impact on report rate.

Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_state_pm.c:18]: I could make a check in release configuration

**Comment by MarekPieta** [comment]
Please keep in mind to rebase and update configurations of nrf5340 DK to prevent CI issues:
`configuration/nrf5340dk_nrf5340_cpuapp/app_ZDebugB0.conf:37: warning: attempt to assign the value 'y' to the undefined symbol DESKTOP_POWER_MANAGER_ENABLE`

File: include/caf/power_manager.rst
Comment (rakons) [reply include/caf/power_manager.rst:1]: Witch notes still references to nrf_desktop? I have done some cleanup... maybe I missed something.

File: subsys/caf/modules/power_manager.c
Comment (rakons) [reply subsys/caf/modules/power_manager.c:328]: Not all the applications uses keep_alive_event - I would disable it by default.

File: subsys/caf/modules/Kconfig.ble_state
Comment (rakons) [reply subsys/caf/modules/Kconfig.ble_state:26]: ok

File: subsys/caf/events/force_power_down_event.c
Comment (rakons) [reply subsys/caf/events/force_power_down_event.c:25]: ok

File: include/caf/events/force_power_down_event.h
Comment (rakons) [reply include/caf/events/force_power_down_event.h:16]: How to generate it?

File: include/caf/power_manager.rst
Comment (MarekPieta) [reply include/caf/power_manager.rst:1]: `If the device is not connected through USB, the module counts time elapsed since the last user interaction (that is, since the last HID report sent from the device).`
We should mention `keep_alive_event` here instead

File: subsys/caf/modules/power_manager.c
Comment (MarekPieta) [reply subsys/caf/modules/power_manager.c:328]: ok

File: include/caf/events/force_power_down_event.h
Comment (MarekPieta) [reply include/caf/events/force_power_down_event.h:16]: Add proper doxygengroup to` nrf/include/caf_overview.rst`

File: include/caf/power_manager.rst
Comment (MarekPieta) [inline include/caf/power_manager.rst:62]: This is actually nRF Desktop specific too

File: applications/nrf_desktop/src/modules/ble_bond.c
Comment (MarekPieta) [inline applications/nrf_desktop/src/modules/ble_bond.c:1037]: `Subscribe` for `power_down_event` and `wake_up_event` only if `IS_ENABLED(CONFIG_CAF_POWER_MANAGER)`

**Review by MarekPieta — APPROVED** [review]

File: applications/nrf_desktop/src/modules/ble_bond.c
Comment (rakons) [reply applications/nrf_desktop/src/modules/ble_bond.c:1037]: Ok

File: include/caf/power_manager.rst
Comment (rakons) [reply include/caf/power_manager.rst:62]: ok

File: subsys/caf/events/Kconfig.keep_alive_event
Comment (tejlmand) [inline subsys/caf/events/Kconfig.keep_alive_event:7]: We have just introduced Pelion client in v1.6.0, and now we are renaming Kconfig symbols which will require all those Kconfig settings to be renamed for end-users if they have made there own configs.

Could we potentially just deprecate the old names for one more release, and have a direct translation, like:
```
config PELION_CLIENT_KEEP_ALIVE_EVENTS
	bool "Enable keep alive events (deprecated)"
	select CAF_KEEP_ALIVE_EVENTS
	help
	  Enable keep alive events.
	  This setting is deprecated, use CAF_KEEP_ALIVE_EVENTS instead.

```

**Review by tejlmand — CHANGES_REQUESTED** [review]

File: subsys/caf/events/Kconfig.keep_alive_event
Comment (tejlmand) [reply subsys/caf/events/Kconfig.keep_alive_event:7]: @Rallare and @OeySa what is your opinion on this ?

File: subsys/caf/modules/Kconfig.ble_state
Comment (tejlmand) [inline subsys/caf/modules/Kconfig.ble_state:27]: And what will happen afterwards ?
Will the system then power down when the current active connection terminates, or will the code need to request a new power down later, if this setting was enabled and a power down was requested.

I think this help text could be extended with more info.

File: subsys/caf/events/Kconfig.force_power_down_event
Comment (tejlmand) [inline subsys/caf/events/Kconfig.force_power_down_event:8]: so we now have a "Force power down` setting, and in the commit before, there was a `This enables small module that blocks power down if we have active BLE connection.`

So who will win ? The force power down, or will it actually wait until all connection are closed, before powering down.

File: applications/nrf_desktop/src/modules/Kconfig.power_manager
Comment (tejlmand) [inline applications/nrf_desktop/src/modules/Kconfig.power_manager:16]: This setting has been there for 3 years: https://github.com/nrfconnect/sdk-nrf/commit/4dbfda77cf20c509fe8fbab298803bb1cd7b1de4
i.e. since release 0.1.0.

Is it really a good idea to just remove this setting without any deprecation first ?

@Rallare and @OeySa feel free to give opinions on this.

**Review by tejlmand — CHANGES_REQUESTED** [review]
I believe we should deprecate settings first, to allow time for our users to adopt, before simply remove settings.

File: subsys/caf/modules/Kconfig.ble_state
Comment (pdunaj) [reply subsys/caf/modules/Kconfig.ble_state:27]: The problem with putting a comprehensive help everywhere is that we would need to maintain copies of it in multiple places. We have help for power manager in rst.

File: applications/nrf_desktop/src/modules/Kconfig.power_manager
Comment (pdunaj) [reply applications/nrf_desktop/src/modules/Kconfig.power_manager:16]: We had cases where options were migrated to CAF. We could deprecate I think - essentially leave this option and select one in caf. I am however unsure if this will really be huge problem to fix during update.
What I mean there is no code that depends on it - there is no associated API. So the only thing to do is to look at diff and accept the change.

File: subsys/caf/modules/Kconfig.ble_state
Comment (tejlmand) [reply subsys/caf/modules/Kconfig.ble_state:27]: i'm not asking about a ten pages documentation, but I think in general we should be better at providing a bit more info.
I feel just a single line like this here is too cheap.

A lot developers don't turn to official documentation until something is not working (myself included).

Besides, what do you mean by maintaining copies ?

Even a re-write of ths docs / user guide will not need the help text in Kconfig to be changed.
A changed Kconfig help text would most likely mean a new setting.

Of course you are right that some of the info will be present in both places, but a re-write of the docs would not mean a change in the Kconfig help, cause the doc is a more complete description with elaborated use cases, where the Kconfig help text would be very specific to that setting, and direct dependencies.

Not an elaborated use-case.

File: applications/nrf_desktop/src/modules/Kconfig.power_manager
Comment (tejlmand) [reply applications/nrf_desktop/src/modules/Kconfig.power_manager:16]: No, but we could have getting started guides, blog posts, etc. that refers to the existing application and settings.

All of a sudden, all such additional information is void, and will cause confusion for the reader.

Comment (tejlmand) [reply applications/nrf_desktop/src/modules/Kconfig.power_manager:16]: also, tech support might be use to asking, have you enabled `XYZ` if something is not working for the user, and suddenly this setting is no longer there.

We should really improve how we rename things, and see it from other people / users perspective.

File: subsys/caf/events/Kconfig.keep_alive_event
Comment (rakons) [reply subsys/caf/events/Kconfig.keep_alive_event:7]: I was trying to provide some solution that would generate warning if deprecated option is selected.
I really like how it is done in u-boot: https://github.com/openil/u-boot/blob/master/Kconfig, and then any deprecated option will depend on this setting. But this is HARD depreciation info, we wish to have a kind of SOFT information.

We can also generate a warning from code if we found that old option is selected.
I get your points, I will implement it in separate commit to allow just reverting to remove it.

But I believe we could implement NRF_DEPRECATED option in our repository and just depend on it all our options that we wish to remove and spend more time to provide good migration manual instead of pleasing the clients that are using NCS 1.6.0 to limit the number of changes they need to do migrating to NCS 1.7.0. It would make sense only if NCS 1.6.0 is LTS. Then we should keep all the changes until the next LTS.

The problem is I did not find any option to make a warning on Kconfig level. If we would just leave it the clients that are using NCS 1.6.0 would never see the warning and they would be surprised just in next release and we spend time and energy on something that finally gives no big gain to nobody. Good migration guide makes the migration in a form of option rename a matter of minutes or seconds.

File: subsys/caf/modules/Kconfig.ble_state
Comment (rakons) [reply subsys/caf/modules/Kconfig.ble_state:27]: There is an error in the option that comes directly from nrf_desktop. Now keep_alive has more general meaning and I have to reword the documentation string here.

Comment (rakons) [reply subsys/caf/modules/Kconfig.ble_state:27]: "This enables keep alive event. This event is processed by power manager and reschedule power down timer. It works only if power manager is in IDLE state. It cannot wake up the device from any sleep states."

Comment (rakons) [reply subsys/caf/modules/Kconfig.ble_state:27]: Ok - at first look I thought it relates to keep alive event - but I like the documentation above to the keep alive event itself - much better than current one.

Comment (rakons) [reply subsys/caf/modules/Kconfig.ble_state:27]: Ok - I have to admit that this decription is not valid. Basically - power manager can still go to power down mode. But the deepest mode would be limited to SUSPENDED if there is active connection - so no power off mode is allowed. I will rephrase it to:
This enables small module that keeps track on active connections.
If there is any active connection the power down modes would be limited to SUSPEND.
Full POWER OFF mode would be only allowed if we have no active connection.

File: subsys/caf/events/Kconfig.force_power_down_event
Comment (rakons) [reply subsys/caf/events/Kconfig.force_power_down_event:8]: This may be hard to understand. I have provided very descriptive help message that should leave no doubts:

	  Enable power manager event to put the device into power down mode when processed.
	  It means the power down would be processed without the delay that is
	  normally applied inside power manager.
	  The event would be send only if the power manager is in IDLE state.
	  The power down event would be send ignoring the fact if any of the modules
	  requires ACTIVE state.
	  The reasult of such situation would be powering down the device to SUSPENDED state.
	  To avoid this the user is responsible in clearing all power ACTIVE requirements
	  before sending this event.
	  The event is here to support very specyfic requirement when the device
	  has to limit is power usage no matter what.
	  The possible usage scenarios are:
	  - Reaction to USB SUSPEND on bus powered device
	  - Reaction to power loss if we have for example to disable LEDs to keep power
	    do to some important jobs before we lost all the power stored in system.

File: applications/nrf_desktop/src/modules/Kconfig.power_manager
Comment (rakons) [reply applications/nrf_desktop/src/modules/Kconfig.power_manager:16]: Then it would be much better to have just a file with:

    config DESKTOP_POWER_MANAGER_ENABLE
    bool "Enable power management"
    depends NCS_DEPRECATED
    help
      This option is deprecated. Use CAF_POWER_MANAGER instead.

**Comment by rakons** [comment]
I have updated the help documentation mentioned above. Lets agree on that how we wish to deprecate old options before I update it.

**Review by pdunaj — APPROVED** [review]

**Comment by pdunaj** [comment]
Hi @tejlmand , there is nothing special in the option mentioned. We had change similar options in the past and heard no complains about it. I don't get why we should since this is simple text replace without any logic.
I don't see why this one should block the PR.

**Comment by pdunaj** [comment]
Hi @greg-fer , can you have a look here since there is doc-required label.

**Comment by nvlsianpu** [comment]
Pleas fix the documentation build.

**Review by greg-fer — APPROVED** [review]
Unfortunately, I don't have much time for more thourough review and preparation of state diagram, so this will be for another task. Approving the current docs after squashing my edits.

**Comment by rakons** [comment]
@greg-fer 
> Approving the current docs after squashing my edits.

You make it harder by merging your changes with my last commit.

**Comment by MarcusCHZ** [comment]
I think this change it okay for me. Moving Power manager module to the common directory make it more easy to use in other apps.

File: applications/nrf_desktop/src/modules/Kconfig.power_manager
Comment (Rallare) [reply applications/nrf_desktop/src/modules/Kconfig.power_manager:16]: Renaming and changing kconfig entries will, in general, force a active change for developers that are updating their own project; thus forcing them to actively have to port their own project. This is something we want to avoid in general, or atleast kept at a backwards compatible level.

In this case is for a specific application, so as long as this change is well documented/handled I do not see any problem. I think something like you mention is perfectly fine when added to the applications own kconfig file, @rakons

**Review by Rallare — APPROVED** [review]

**Comment by pdunaj** [comment]
Super. Thanks for your input. Let me rebase the PR to get rid of the conflicts.

**Comment by pdunaj** [comment]
Rebased

**Comment by pdunaj** [comment]
I run a build locally and we have a small overflow on one of the platforms with small flash. Let me fix that one before merging.

**Comment by pdunaj** [comment]
Done - there was a few byte overflow on nrf52820 dongle. Disabled printk so we should have enough free space for some time.

**Review by tejlmand — APPROVED** [review]

**Pull request merged**
